File: backend/app/models/release.py
import uuid

from sqlalchemy import UUID, Column, DateTime, String
from sqlalchemy.orm import Mapped, relationship
from sqlalchemy.sql import func

from app.db import Base


class Release(Base):
  __tablename__ = "release"

  id: Mapped[UUID] = Column(UUID, primary_key=True, index=True, default=uuid.uuid4)

  name: Mapped[str] = Column(String)
  short: Mapped[str] = Column(String)
  notes: Mapped[str] = Column(String)
  thumb: Mapped[str] = Column(String)
  year: Mapped[int] = Column(String)
  genre: Mapped[str] = Column(String)
  styles: Mapped[str] = Column(String)
  format: Mapped[str] = Column(String)
  purchased_at: Mapped[str] = Column(String)

  release_artists = relationship("ArtistReleaseAssociation", back_populates="release")
  release_labels = relationship("LabelReleaseAssociation", back_populates="release")

  tracks = relationship("Track", back_populates="release")

  created_at: Mapped[DateTime] = Column(
    DateTime(timezone=True), server_default=func.now()
  )
  updated_at: Mapped[DateTime] = Column(DateTime(timezone=True), onupdate=func.now())

File: frontend/src/models/Release.ts
export interface Release {
  [key: string]: any;
  id: string;
  short: string;
  year: string;
  format: string;
  genre: string;
  purchased_at: string;
  styles: string;
  thumb: string;
  created_at: string;
  updated_at: string | null;
  name: string;
  notes: string;
  tracks: Track[];
  labels: Label[];
  artists: Artist[];
}

export interface Track {
  rating: number | null;
  created_at?: string;
  updated_at?: string | null;
  name: string;
  key: string;
  bpm: number;
  genre: string;
  audio: Blob | string;
  length: number;
  id: string;
  side: string;
  release_id: string;
}

interface Label {
  id: string;
  name: string;
  created_at: string;
  release_id: string;
  updated_at: string | null;
}

interface Artist {
  name: string;
  created_at: string;
  id: string;
  release_id: string;
  updated_at: string | null;
}
